packages/amplify-category-api/src/provider-utils/awscloudformation/PipelineWithAwaiter.ts

```diff
@@ -0,0 +1,380 @@
+        ],
+      },
+    ]);
+
+    const role = getRole(scope, `Pipeline`, new iam.ServicePrincipal('codepipeline.amazonaws.com'));
+
+    const pipeline = new codepipeline.Pipeline(scope, `${id}Pipeline`, {
+      pipelineName: `codepipeline-amplify-${service.serviceName}`,
+      crossAccountKeys: false,
+      artifactBucket: bucket,
+      stages: stagesWithDeploy,
+      role,
+    });
+
+    pipeline.node.addDependency(service);
+
+    const pipelineAwaiter = new PipelineAwaiter(scope, 'Awaiter', {
+      pipeline,
+      artifactBucketName: bucket.bucketName,
+      artifactKey: s3SourceActionKey,
+      deploymentMechanism,
+    });
+  }
+}
+
+function createPreBuildStages(
+  scope: cdk.Construct,
+  {
+    bucket,
+    s3SourceActionKey,
+    githubSourceActionInfo,
+    sourceOutput,
+    roleName,
+  }: {
+    bucket: s3.IBucket;
+    s3SourceActionKey: string;
+    githubSourceActionInfo?: GitHubSourceActionInfo;
+    sourceOutput: codepipeline.Artifact;
+    roleName: string;
+  },
+) {
+  const stages: codepipeline.StageOptions[] = [];
+
+  const stage = {
+    stageName: 'Source',
+    actions: [],
+  };
+
+  stages.push(stage);
+
+  if (githubSourceActionInfo && githubSourceActionInfo.path) {
+    const { path, tokenSecretArn } = githubSourceActionInfo;
+    const { owner, repo, branch } = getGitHubOwnerRepoFromPath(path);
+
+    const preBuildOutput = new codepipeline.Artifact('PreBuildArtifact');
+
+    stage.actions = [
+      new codepipelineactions.GitHubSourceAction({
+        actionName: 'Source',
+        oauthToken: cdk.SecretValue.secretsManager(tokenSecretArn),
+        owner,
+        repo,
+        branch,
+        output: preBuildOutput,
+      }),
+    ];
+
+    stages.push({
+      stageName: 'PreBuild',
+      actions: [
+        new codepipelineactions.LambdaInvokeAction({
+          actionName: 'PreBuild',
+          lambda: new lambda.Function(scope, 'PreBuildLambda', {
+            code: lambda.S3Code.fromBucket(bucket, 'codepipeline-action-buildspec-generator-lambda.zip'),
+            handler: 'index.handler',
+            runtime: lambda.Runtime.NODEJS_12_X,
+          }),
+          inputs: [preBuildOutput],
+          outputs: [sourceOutput],
+        }),
+      ],
+    });
+  } else {
+    stage.actions = [
+      new codepipelineactions.S3SourceAction({
+        role: getRole(scope, roleName),
+        actionName: 'Source',
+        bucket,
+        bucketKey: s3SourceActionKey,
+        output: sourceOutput,
+      }),
+    ];
+  }
+
+  return stages;
+}
+
+function getRole(scope: cdk.Construct, prefix: string, assumedBy?: iam.IPrincipal): iam.Role {
+  const role = new iam.Role(scope, `${prefix}Role`, {
+    assumedBy: assumedBy ?? new iam.AccountRootPrincipal(),
+  });
+
+  const cfnRole = role.node.defaultChild as iam.CfnRole;
+
+  // We add a dummy statement that we immediately remove so CDK creates a policy to which we can add a condition
+  const defaultPolicy = role.addToPrincipalPolicy(
```

Comment:
What's the reasoning for this DENY policy?

Reply:
Removed this